compiler/src/dotty/tools/dotc/interactive/Interactive.scala

```diff
@@ -165,38 +312,91 @@ object Interactive {
-        // FIXME: We shouldn't need a cast. Change NavigateAST.pathTo to return a List of Tree?
-        val path = NavigateAST.pathTo(pos.pos, tree.tree, skipZeroExtent = true).asInstanceOf[List[untpd.Tree]]
+      case Some(tree) => pathTo(tree.tree, pos.pos)
+      case None => Nil
+    }
 
-        path.dropWhile(!_.hasType).asInstanceOf[List[tpd.Tree]]
-      case None =>
-        Nil
+  def pathTo(tree: Tree, pos: Position)(implicit ctx: Context): List[Tree] =
+    if (tree.pos.contains(pos)) {
+      // FIXME: We shouldn't need a cast. Change NavigateAST.pathTo to return a List of Tree?
+      val path = NavigateAST.pathTo(pos, tree, skipZeroExtent = true).asInstanceOf[List[untpd.Tree]]
+      path.dropWhile(!_.hasType).asInstanceOf[List[tpd.Tree]]
     }
+    else Nil
+
+  def contextOfStat(stats: List[Tree], stat: Tree, exprOwner: Symbol, ctx: Context): Context = stats match {
+    case Nil =>
+      ctx
+    case first :: _ if first eq stat =>
+      ctx.exprContext(stat, exprOwner)
+    case (imp: Import) :: rest =>
+      contextOfStat(rest, stat, exprOwner, ctx.importContext(imp, imp.symbol(ctx)))
+    case _ =>
+      ctx
+  }
+
+  def contextOfPath(path: List[Tree])(implicit ctx: Context): Context = path match {
+    case Nil | _ :: Nil =>
+      ctx.run.runContext.fresh.setCompilationUnit(ctx.compilationUnit)
+    case nested :: encl :: rest =>
+      import typer.Typer._
+      val outer = contextOfPath(encl :: rest)
+      encl match {
+        case tree @ PackageDef(pkg, stats) =>
+          assert(tree.symbol.exists)
+          if (nested `eq` pkg) outer
+          else contextOfStat(stats, nested, pkg.symbol.moduleClass, outer.packageContext(tree, tree.symbol))
+        case tree: DefDef =>
+          assert(tree.symbol.exists)
+          val localCtx = outer.localContext(tree, tree.symbol).setNewScope
+          for (tparam <- tree.tparams) localCtx.enter(tparam.symbol)
+          for (vparams <- tree.vparamss; vparam <- vparams) localCtx.enter(vparam.symbol)
+            // Note: this overapproximates visibility a bit, since value parameters are only visible
+            // in subsequent parameter sections
+          localCtx
+        case tree: MemberDef =>
+          assert(tree.symbol.exists)
+          outer.localContext(tree, tree.symbol)
+        case tree @ Block(stats, expr) =>
+          val localCtx = outer.fresh.setNewScope
+          stats.foreach {
+            case stat: MemberDef => localCtx.enter(stat.symbol)
+            case _ =>
+          }
+          contextOfStat(stats, nested, ctx.owner, localCtx)
+        case tree @ CaseDef(pat, guard, rhs) if nested `eq` rhs =>
+          val localCtx = outer.fresh.setNewScope
+          pat.foreachSubTree {
+            case bind: Bind => localCtx.enter(bind.symbol)
+            case _ =>
+          }
+          localCtx
+        case tree @ Template(constr, parents, self, _) =>
+          if ((constr :: self :: parents).exists(nested `eq` _)) ctx
```

Comment:
Any reason to not use `.contains(nested)` here?

Reply:
None. I overlooked that tree comparison is by `eq` anyway.